### PR TITLE
[Feature] Approval group

### PR DESCRIPTION
**Pull request by EliasBoulharts**

### CODE REVIEW

File: profiles/apps.py [hunk 1]
```diff
@@ -1,13 +1,19 @@
...
+
 
 def create_roles(sender, **kwargs):
+    logger.info("create_roles method called")
     from django.contrib.auth.models import Permission
     from profiles.models import Role
     from django.contrib.contenttypes.models import ContentType
     from profiles.models.rbac import roles_config
     for model_name, roles in roles_config.items():
+        print(model_name)
```
Comment (a-belhadj) [inline profiles/apps.py:16]: to be removed

Comment (a-belhadj) [inline profiles/apps.py:16]: to be removed

File: service_catalog/models/approval_step.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        blank=False,
+        null=False,
+    )
+    next = ForeignKey(
+        "service_catalog.ApprovalStep",
+        blank=True,
+        null=True,
+        default=None,
+        on_delete=SET_NULL,
+        related_name="previous_list"
```
Comment (a-belhadj) [inline service_catalog/models/approval_step.py:46]: previous

Comment (a-belhadj) [inline service_catalog/models/approval_step.py:46]: previous

File: service_catalog/models/approval_step.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        elif self.type == ApprovalStepType.ALL_OF_THEM:
+            if state_list.count(ApprovalState.APPROVED) == len(state_list):
+                return ApprovalState.APPROVED
+            else:
+                return ApprovalState.PENDING
+        elif self.type == ApprovalStepType.AT_LEAST_ONE:
+            if ApprovalState.APPROVED in state_list:
+                return ApprovalState.APPROVED
+            else:
+                return ApprovalState.PENDING
```
Comment (a-belhadj) [inline service_catalog/models/approval_step.py:103]: unit test needed

Comment (a-belhadj) [inline service_catalog/models/approval_step.py:103]: unit test needed

File: service_catalog/models/approval_workflow.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    )
+
+    def __str__(self):
+        return self.name
+
+    def clean(self):
+        if self.entry_point and self.entry_point.approval_workflow != self:
+            raise ValidationError({'entry_point': _("The entry point must be in the workflow.")})
+
+    def update_position(self):
```
Comment (a-belhadj) [inline service_catalog/models/approval_workflow.py:27]: with an s + comments

File: service_catalog/models/request.py [hunk 5]
```diff
@@ -27,23 +32,44 @@
...
+    date_archived = DateTimeField(auto_now=False, blank=True, null=True)
+    tower_job_id = IntegerField(blank=True, null=True)
     state = FSMField(default=RequestState.SUBMITTED, choices=RequestState.choices)
-    periodic_task = models.ForeignKey(PeriodicTask, on_delete=models.SET_NULL, null=True, blank=True)
-    periodic_task_date_expire = models.DateTimeField(auto_now=False, blank=True, null=True)
-    failure_message = models.TextField(blank=True, null=True)
+    periodic_task = ForeignKey(PeriodicTask, on_delete=SET_NULL, null=True, blank=True)
+    periodic_task_date_expire = DateTimeField(auto_now=False, blank=True, null=True)
+    failure_message = TextField(blank=True, null=True)
+    approval_step = ForeignKey(
```
Comment (a-belhadj) [inline service_catalog/models/request.py:48]: what if an approval workflow is deleted before the request is accepted ?

Comment (a-belhadj) [inline service_catalog/models/request.py:48]: what if an approval workflow is deleted before the request is accepted ?

File: service_catalog/models/approval_step.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    """
+    if created:
+        for team in instance.teams.all():
+            instance.add_team_in_role(team, "Approver")
+        if not instance.approval_workflow.entry_point or instance.next == instance.approval_workflow.entry_point:
+            instance.approval_workflow.entry_point = instance
+            instance.approval_workflow.save()
+
+
+def teams_changed(sender, instance, **kwargs):
```
Comment (Sispheor) [inline service_catalog/models/approval_step.py:152]: add a test

File: service_catalog/models/approval_step_state.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
+from django.db.models import Model, ForeignKey, CASCADE, CharField
+from service_catalog.models.approval_state import ApprovalState
+
+
+class ApprovalStepState(Model):
```
Comment (Sispheor) [inline service_catalog/models/approval_step_state.py:5]: add user (the user who has validated in the team)

File: profiles/apps.py (hunk 1 above)
Comment (EliasBoulharts) [reply profiles/apps.py]: done

Comment (EliasBoulharts) [reply profiles/apps.py]: done

File: service_catalog/models/approval_step.py (hunk 2 above)
Comment (EliasBoulharts) [reply service_catalog/models/approval_step.py]: done

File: service_catalog/models/request.py [hunk 8]
```diff
@@ -27,23 +32,44 @@
...
+        null=True,
+        on_delete=SET_NULL,
+        related_name='requests',
+        related_query_name='request'
+    )
 
     def __str__(self):
         return f"{self.operation.name} - {self.instance.name} (#{self.id})"
 
+    def get_state(self, user=None):
```
Comment (Sispheor) [inline service_catalog/models/request.py:60]: to be removed

File: service_catalog/models/request.py [hunk 9]
```diff
@@ -225,6 +240,66 @@ def check_job_status(self):
             self.periodic_task.delete()
             send_mail_request_update(target_request=self)
 
+    def get_state_from_approval_step(self, user, target_approval_step_state):
+        """
+        This methode get status of all
```
Comment (Sispheor) [inline service_catalog/models/request.py:245]: method

File: service_catalog/views/approval_step_views.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+@user_passes_test(lambda u: u.is_superuser)
+def approval_step_graph(request, approval_workflow_id):
+    approval_workflow = get_object_or_404(ApprovalWorkflow, id=approval_workflow_id)
+    if not approval_workflow.entry_point:
+        return redirect('service_catalog:approval_step_create', approval_workflow_id)
+    approval_step_list = list()
+    current = approval_workflow.entry_point
+    while current:
+        approval_step_list.append(current)
+        current = current.next
```
Comment (Sispheor) [inline service_catalog/views/approval_step_views.py:106]: use position instead

File: tests/utils.py [hunk 11]
```diff
@@ -5,4 +5,5 @@ def check_data_in_dict(self, expected_data_list, data_list):
             if isinstance(val_var, list) and isinstance(data[key_var], list):
                 self.assertEqual(set(val_var), set(data[key_var]))
             else:
+                print(key_var)
```
Comment (Sispheor) [inline tests/utils.py:8]: to be removed

**Review by Sispheor — APPROVED** [review]

**Pull request merged**